ext/liquid_c/block.c

```diff
@@ -123,65 +210,199 @@ static VALUE internal_block_parse(VALUE self, VALUE tokens, VALUE parse_context,
+                if (body->blank && !RTEST(rb_funcall(new_tag, intern_is_blank, 0)))
+                    body->blank = false;
 
-                if (rb_ivar_get(self, intern_blank) == Qtrue && !RTEST(rb_funcall(new_tag, intern_is_blank, 0)))
-                    rb_ivar_set(self, intern_blank, Qfalse);
+                block_body_add_write_node(body, new_tag);
+                render_score_increment += 1;
+                break;
+            }
+        }
+    }
+loop_break:
+    body->render_score += render_score_increment;
+    return unknown_tag;
+}
+
+static void ensure_not_parsing(block_body_t *body)
+{
+    if (body->parsing) {
+        rb_raise(rb_eRuntimeError, "Liquid::C::BlockBody is in a incompletely parsed state");
+    }
+}
+
+static VALUE block_body_parse(VALUE self, VALUE tokenizer_obj, VALUE parse_context_obj)
+{
+    parse_context_t parse_context = {
+        .tokenizer_obj = tokenizer_obj,
+        .ruby_obj = parse_context_obj,
+    };
+    Tokenizer_Get_Struct(tokenizer_obj, parse_context.tokenizer);
+    block_body_t *body;
+    BlockBody_Get_Struct(self, body);
+
+    ensure_not_parsing(body);
+    if (body->source == Qnil) {
+        body->source = parse_context.tokenizer->source;
+    } else if (body->source != parse_context.tokenizer->source) {
+        rb_raise(rb_eArgError, "Liquid::C::BlockBody#parse must be passed the same tokenizer when called multiple times");
+    }
+    body->parsing = true;
+    // remove terminating OP_LEAVE to resume parsing
+    body->instructions.size -= 1;
+
+    tag_markup_t unknown_tag = internal_block_body_parse(body, &parse_context);
+    block_body_add_leave(body);
+    body->parsing = false;
+    return rb_yield_values(2, unknown_tag.name, unknown_tag.markup);
+}
+
+static VALUE block_body_render_to_output_buffer(VALUE self, VALUE context, VALUE output)
+{
+    block_body_t *body;
+    BlockBody_Get_Struct(self, body);
+    liquid_vm_render(body, context, output);
+    return output;
+}
 
-                rb_ary_push(nodelist, new_tag);
+static VALUE block_body_blank_p(VALUE self)
+{
+    block_body_t *body;
+    BlockBody_Get_Struct(self, body);
+    return body->blank ? Qtrue : Qfalse;
+}
+
+static VALUE block_body_remove_blank_strings(VALUE self)
+{
+    block_body_t *body;
+    BlockBody_Get_Struct(self, body);
+
+    if (!body->blank) {
+        rb_raise(rb_eRuntimeError, "remove_blank_strings only support being called on a blank block body");
+    }
+    ensure_not_parsing(body);
+
+    size_t *const_ptr = (size_t *)body->constants.data;
+    const uint8_t *ip = body->instructions.data;
+
+    while (true) {
+        switch (*ip++) {
+            case OP_LEAVE:
+                goto loop_break;
+            case OP_WRITE_RAW:
+            {
+                const_ptr++;
+                size_t *size_ptr = const_ptr++;
+                if (*size_ptr) {
+                    *size_ptr = 0; // effectively a no-op
+                    body->render_score--;
+                }
```

Comment:
As mentioned in https://github.com/Shopify/liquid-c/pull/59#discussion_r490293174

> I'm not concerned about further optimizing this edge case. Primarily these blank strings are handled in this way to minimize impact on code that isn't relying on this feature while keeping backwards compatibility,

Specifically, once we have jump instructions for control flow tags, we would also be forced to update all the affected jump offsets.  That will be easier to introduce without having to deal with jump offset relocation, but afterwards an extra pass could be added that is more clearly beneficial, since we could replace pessimistically large forward jump instructions (e.g. a decoded 24-bit offset or aligned 32-bit offset) with smaller jump instructions (e.g. a byte offset) where possible.